bumble/hid.py

```diff
@@ -242,37 +297,224 @@ def on_l2cap_channel_open(self, l2cap_channel: l2cap.ClassicChannel) -> None:
+        SUCCESS = 0xFF
+
+    class GetSetStatus:
+        def __init__(self) -> None:
+            self.data: bytes
+            self.status = 0
+
+    def __init__(self, device: Device) -> None:
+        super().__init__(device, HID.Role.DEVICE)
+        self.get_report_cb = None
+        self.set_report_cb = None
+        self.get_protocol_cb = None
+        self.set_protocol_cb = None
+
+    def send_handshake_message(self, result_code: int) -> None:
+        msg = SendHandshakeMessage(result_code)
+        hid_message = bytes(msg)
+        logger.debug(f'>>> HID HANDSHAKE MESSAGE, PDU: {hid_message.hex()}')
+        self.send_pdu_on_ctrl(hid_message)
+
+    def send_control_data(self, report_type: int, data: bytes):
+        msg = SendControlData(report_type=report_type, data=data)
+        hid_message = bytes(msg)
+        logger.debug(f'>>> HID CONTROL DATA: {hid_message.hex()}')
+        self.send_pdu_on_ctrl(hid_message)
+
+    def handle_get_report(self, pdu: bytes):
+        ret = self.GetSetStatus()
+        report_type = pdu[0] & 0x03
+        buffer_flag = (pdu[0] & 0x08) >> 3
+        report_id = pdu[1]
+        logger.debug("buffer_flag: " + str(buffer_flag))
+        if buffer_flag == 1:
+            buffer_size = (pdu[3] << 8) | pdu[2]
+        else:
+            buffer_size = 0
+
+        if self.get_report_cb != None:
+            ret = self.get_report_cb(
+                report_id, report_type, buffer_size
+            )  #  type:  ignore
+
+            if ret.status == self.GetSetReturn.FAILURE:
+                self.send_handshake_message(Message.Handshake.ERR_UNKNOWN)
+            elif ret.status == self.GetSetReturn.SUCCESS:
+                data = bytearray()
+                data.append(report_id)
+                data.extend(ret.data)
+                if (
+                    len(data) < self.l2cap_ctrl_channel.mtu
+                ):  #  type:  ignore[union-attr]
+                    self.send_control_data(report_type=report_type, data=data)
+                else:
+                    self.send_handshake_message(Message.Handshake.ERR_INVALID_PARAMETER)
+
+            elif ret.status == self.GetSetReturn.REPORT_ID_NOT_FOUND:
+                self.send_handshake_message(Message.Handshake.ERR_INVALID_REPORT_ID)
+            elif ret.status == self.GetSetReturn.ERR_INVALID_PARAMETER:
+                self.send_handshake_message(Message.Handshake.ERR_INVALID_PARAMETER)
+            elif ret.status == self.GetSetReturn.ERR_UNSUPPORTED_REQUEST:
+                self.send_handshake_message(Message.Handshake.ERR_UNSUPPORTED_REQUEST)
+        else:
+            logger.debug("GetReport callback not registered !!")
+            self.send_handshake_message(Message.Handshake.ERR_UNSUPPORTED_REQUEST)
+
+    def register_get_report_cb(self, cb):
+        self.get_report_cb = cb
+        logger.debug("GetReport callback registered successfully")
+
+    def handle_set_report(self, pdu: bytes):
+        if self.set_report_cb != None:
```

Comment:
See earlier comment about testing for None at the start and returning.

Reply:
Resolved